src/Kiota.Builder/KiotaBuilder.cs

```diff
@@ -1583,6 +1584,76 @@ private CodeClass AddModelClass(OpenApiUrlTreeNode currentNode, OpenApiSchema sc
                 .Where(static x => x.Value != null)
                 .Select(static x => KeyValuePair.Create(x.Key, x.Value!));
     }
+    private static IEnumerable<CodeElement> GetAllModels(CodeNamespace currentNamespace)
+    {
+        return currentNamespace.Classes
+                            .Where(static x => x.IsOfKind(CodeClassKind.Model))
+                            .OfType<CodeElement>()
+                            .Union(currentNamespace.Enums)
+                            .Union(currentNamespace.Namespaces.SelectMany(static x => GetAllModels(x)));
+    }
+    private void TrimInheritedModels()
+    {
+        if (modelsNamespace is null || rootNamespace is null) return;
+        var models = GetAllModels(modelsNamespace);
+        var allModelsInUse = GetTypeDefinitionsInNamespace(rootNamespace).ToHashSet();
+        var allRelatedModels = allModelsInUse.SelectMany(x => GetRelatedClasses(x)).ToHashSet(); // so base classes and properties definitions are also included
+        Parallel.ForEach(models, x =>
+        {
+            if (allRelatedModels.Contains(x) || allModelsInUse.Contains(x)) return;
+            if (x is CodeClass currentClass)
+            {
+                var baseClass = currentClass.GetParentClass();
+                if (baseClass != null && allModelsInUse.Contains(baseClass))
+                    return;
+                if (baseClass != null)
+                    do
+                    {
+                        baseClass.DiscriminatorInformation.RemoveDiscriminatorMapping(currentClass);
+                    } while ((baseClass = baseClass.GetParentClass()) != null); // discriminator might also be in grand parent types
+            }
+            x.GetImmediateParentOfType<CodeNamespace>().RemoveChildElement(x);
+        });
+    }
+    private IEnumerable<CodeElement> GetRelatedClasses(CodeElement currentElement, HashSet<CodeElement>? visited = null)
+    {
+        if (currentElement is not CodeClass currentClass)
+            return Enumerable.Empty<CodeElement>();
+        visited ??= new();
+        var propertiesDefinitions = currentClass.Properties
+                            .SelectMany(static x => x.Type.AllTypes)
+                            .Select(static x => x.TypeDefinition)
+                            .Where(static x => x is CodeClass || x is CodeEnum)
+                            .Select(static x => x!)
```

Comment:
```suggestion
                            .Select(static x => x.TypeDefinition!)
                            .Where(static x => x is CodeClass || x is CodeEnum)
```

Reply:
I need the select otherwise the null reference types check complains. It doesn't understand the type assertion makes it impossible to be null. I wished LINQ had of type with multiple type parameters.

Reply:
@baywet. 
I moved the null assertion up to line 1625 as `.Select(static x => x.TypeDefinition!)` motivated by the code line 1653 which does the same thing.

Reply:
ah right, thank you for pointing this out, I missed it